streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBVersionedStoreSegmentValueFormatter.java

```diff
@@ -495,6 +501,41 @@ private boolean isLastIndex(final int index) {
             return unpackedReversedTimestampAndValueSizes.get(index).timestamp == minTimestamp;
         }
 
+        private void truncateRecordsToTimestamp(final long timestamp) {
+            if (timestamp <= minTimestamp) {
+                // delete everything in this current segment by replacing it with a degenerate segment
+                initializeWithRecord(new ValueAndValueSize(null), timestamp, timestamp);
+                return;
+            }
+
+            final SegmentSearchResult searchResult = find(timestamp, false);
+            // all records with later timestamps should be removed
+            int fullRecordsToTruncate = searchResult.index();
+            // additionally remove the current record as well, if its validFrom equals the
+            // timestamp to truncate to
+            if (searchResult.validFrom() == timestamp) {
+                fullRecordsToTruncate++;
+            }
+
+            if (fullRecordsToTruncate == 0) {
+                // no records to remove; update nextTimestamp and return
```

Comment:
This catches the situation in which the last record of the segment being inserted into has `validFrom=5` and `validTo=10`, and `insertAsLatest()` has been called with timestamp `8` (more generally, any timestamp larger than 5 and smaller than 10). In this situation, all we have to do to recover (before performing the insert) is to update `validTo` for the existing record from 10 to 8.

As noted in a comment above, I don't actually expect this situation to be hit but I think it is good to be defensive about these additional edge cases in this recovery logic, as it is difficult to anticipate the types of failures which may be encountered.

Reply:
Well, but if we want to make it generic, could it not be that we actually have two record `[5,10)[10,15)` and try to insert at 8? Or would we never call `insertAsLatest()` for this case?

Reply:
The logic added in this PR does handle this case. (It's not "expected" that this case will come up, but it is possible in weird edge cases such as the examples that you gave above.) There are unit tests for it as well.

Reply:
But if we handle this case, don't we incorrectly drop `[10,15)`? The intention if my example was to say that `[5,10)` was the partial write, but `[10,15)` was a clean write later (due to re-ordering) that did not detect the corruption and went through cleanly, and now inserting `8` would incorrectly purge `[10,15)`?

Or maybe this case could never happen? -- Just want to double check.

Reply:
That can never happen -- when a single `put()` call requires writing to multiple segments, it always writes to the latest record in the older segment and the earliest record in the newer segment. If `[5,10)` is a partial write, then that means it was written as the latest record to an older segment, and that the newer segment already has that same record timestamp 5 in it as its earliest record (which was supposed to be replaced with a new record except that second write failed). In that case, any subsequent write with timestamp greater than 5 will stop in the newer segment because the newer segment still has minTimestamp = 5.